test_quality_of_service/test/qos_utilities.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+
+std::tuple<size_t, size_t> chrono_milliseconds_to_size_t(
+  const std::chrono::milliseconds & milliseconds)
+{
+  size_t seconds = milliseconds.count() / 1000;
+  size_t nanoseconds = (milliseconds.count() % 1000) * 1000000;
+  return std::make_tuple(seconds, nanoseconds);
+}
+
+Publisher::Publisher(
+  const std::string name,
+  const std::string topic,
+  rclcpp::PublisherOptions<> pub_options,
+  std::chrono::milliseconds publish_period)
+: Node(name),
+  name_(name),
+  topic_(topic),
+  pub_options_(pub_options),
+  publish_period_(publish_period)
+{
+  RCLCPP_INFO(this->get_logger(), "created publisher %s %s\n", name.c_str(), topic.c_str());
+  publisher_ = this->create_publisher<std_msgs::msg::String>(topic_, pub_options_);
+}
+
+void Publisher::start_publishing()
+{
+  toggle_publish_mutex.lock();
+
+  if (is_publishing) {
+    return;
+  }
+
+  is_publishing = true;
+
+  auto timer_callback =
+    [this]() -> void {
+      auto message = std_msgs::msg::String();
+
+      message.data = this->name_ + ": testing " + std::to_string(this->sent_message_count_++);
+
+      RCLCPP_INFO(this->get_logger(), this->name_ + ": publishing '%s'", message.data.c_str());
+
+      this->publisher_->publish(message);
+    };
+  timer_ = this->create_wall_timer(publish_period_, timer_callback);
+
+  toggle_publish_mutex.unlock();
+}
+
+void Publisher::stop_publishing()
+{
+  toggle_publish_mutex.lock();
+  is_publishing = false;
+
+  if (timer_) {
+    timer_->cancel();
+  }
+  toggle_publish_mutex.unlock();
+}
+
+void Publisher::toggle_publishing()
```

Comment:
This is why you need the recursive mutex and this is very fragile, instead, consider one of the following options:

A/ refactor the code as follow (of course for real code I'd move the timer_callback into it's own function, etc.):

```c++
void Publisher::toggle_publishing_impl(bool allow_start, bool allow_stop) {
  auto timer_callback =
    [this]() -> void {
      this->sent_message_count_++;
      auto message = std_msgs::msg::String();
      std::ostringstream ss;
      ss << this->name_ << ": testing" << this->sent_message_count_;
      message.data = ss.str();
      RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());
      this->publisher_->publish(message);
    };
  auto timer = this->create_wall_timer(publish_period_, timer_callback);

  {
    std::scoped_lock lock(toggle_publish_mutex_);
    if (is_publishing && allow_stop) {
      return;
    }
    if (!is_publishing && allow_start) {
      return;
    }
    if (is_publishing) {
      timer_->cancel();
      timer_.reset();
    } else {
      timer_ = timer;
    }
    is_publishing = !is_publishing;
  }
}

void Publisher::toggle_publishing() {
   return toggle_publishing_impl(true /* allow_start */, true /* allow_stop */);
}

void Publisher::start_publishing() {
   return toggle_publishing_impl(true /* allow_start */, false /* allow_stop */);
}

void Publisher::start_publishing() {
   return toggle_publishing_impl(false /* allow_start */, true /* allow_stop */);
}
```

B/ rename this `try_toggle_publishing` and cope with the (potential) differences between `is_publishing` state when you read it in toggle and in start by returning a boolean indicating the operation failed.